docs/man/run-as-anonymous.1
.TH run\-as\-anonymous 1   
.SH NAME
run\-as\-anonymous \- Start process with an unused UID, in a chroot() environment
.SH SYNOPSIS
.nf
run\-as\-anonymous \fIcommand\fR \fIargs...\fR
.fi
.SH DESCRIPTION
run\-as\-anonymous is a small setuid\-root program for starting processes
such that they are isolated from the rest of the system. It chooses a
UID for the process that is not being used by other processes. It
sets the process's root directory using chroot().
.PP
run\-as\-anonymous is part of Plash. Plash uses it to give a user
protection from the programs they run.
.PP
Only a small part of Plash, including run\-as\-anonymous, needs root
access to install. If you are a system administrator, this means you
can inspect this part and install it in confidence that it won't
interfere with the rest of your system. You don't have to trust the
rest of Plash, which is more complex \-\- though your users will trust
the rest of Plash if they use it.
.SH "THE CHROOT() JAIL"
The chroot() jail directory is typically \fB/var/lib/plash\-chroot\-jail\fR. It needs to be on a writable filesystem
because it contains a lock directory.
.PP
The jail directory typically contains one directory, \fBspecial\fR, which \fIcan\fR be writable by a normal user without
effectively giving that user root authority. (This is not relevant if
you are installing an RPM or Debian package of Plash.)
.PP
The jail directory should \fInot\fR be writable by a normal
user. If this were so, the user could hard link a setuid\-root
executable into the jail directory, and put a file \fBlib/ld\-linux.so.2\fR containing their own code into the jail directory.
Running this executable using run\-as\-anonymous would run the user's
code with root permission. The user would effectively have root
authority.
.PP
The idea behind the \fBspecial\fR directory is that no
setuid\-root executable will depend on the contents of \fB/special\fR.
.PP
Typically, \fBspecial\fR will only contain Plash's version of
glibc's dynamic linker, \fBld\-linux.so.2\fR. The dynamic
linker is invoked as an executable. It will load a normal executable
from a file outside the chroot jail, via the file descriptors that it
is passed. run\-as\-anonymous does execve() after changing the root
directory, so there needs to be an executable in the chroot jail to
call execve() on.
.SH "ALLOCATING USER IDS"
run\-as\-anonymous allocates UIDs from a numeric range that is specific
as a compile time option (typically 0x100000 to 0x200000) \-\- see
config.sh.
.PP
It creates lock files to ensure that UIDs are not re\-used. The lock
files are typically kept in \fB/var/lib/plash\-chroot\-jail/plash\-uid\-locks\fR.
.PP
The program gc\-uid\-locks can be run every so often to remove lock
files when the corresponding UID is no longer used by any process.
.PP
The file \fB/var/lib/plash\-chroot\-jail/plash\-uid\-locks/flock\-file\fR is locked using
flock() to ensure that gc\-uid\-locks does not run concurrently with any
instance of run\-as\-anonymous.
.SH "AUTHORITY GRANTED AND DENIED"
Putting a process in a chroot() jail can deny it the ability to open
most files. Giving a process a unique UID denies the ability to send
signals to other processes (using kill()) or take over other processes
(using ptrace()).
.PP
A process can communicate with the outside world using the file
descriptors it was created with, including receiving other file
descriptors (via Unix domain sockets). This is how Plash grants a
process authority, after run\-as\-anonymous has taken away authority
that could be exercised through the more usual system calls.
.PP
Note that if a process receives a directory file descriptor, it can
use that to break out of its chroot() jail, using fchdir(). Plash is
careful not to pass a process any directory file descriptors.
.PP
run\-as\-anonymous does not prevent a process from connecting to or
listening on network sockets.
.SH "SEE ALSO"
gc\-uid\-locks, plash
.PP
/usr/share/doc/plash
.SH AUTHOR
Mark Seaborn <mseaborn@onetel.com>


debian/changelog
plash (1.15.1) unstable; urgency=low

  * Add autoconf script (configure.in).
  * add PLASH_LIBC_DEBUG env var
  * pola-shell.c:  Command lists don't have to end with ';'.  Command line now accepts command lists as well as individual commands.
  * exec-object.c:  fix to check for error code from install_fds().
  * libc-comms.c:  plash_libc_reset_connection() now allows close() and dup2() to clobber comm FD.
  * powerbox.c, gtk-powerbox2.c:  Add support for WM_TRANSIENT_FOR hint.
  * Fork gtk-powerbox that inherits from GtkDialog and suppresses the opening of a window by implementing the "map" method.
  * gtk-powerbox.c:  Fix handling of cancel case.
  * filesysobj-real.c:  real_file_socket_connect():  Fixed
    symlink-related race condition.  It now works by hard linking the
    Unix domain socket into a directory in /tmp, so it only works if
    the socket is on the same device as /tmp.
     * Wrote a test case that exploits the old race condition.
  * filesysobj-real.c:  real_file_chmod():  Removed use of chmod() call
    to fix symlink-related race condition.  Now use open() and fchmod().
    Unfortunately, this will not work if read permission is not already
    enabled for the file.
  * filesysobj-real.c:  real_file_utimes():  Fixed symlink-related race
    condition.  Now use open() and futimes() instead of utimes().
  * XXMLParse.pm:  Add >> quoting syntax.
  * pola-run.c:  Set up the PLASH_FAKE_{E,}[UG]ID environment variables
    (as shell.c does).
  * pola-run.c:  Executable filename is now resolved in callee's
    namespace.  This makes sense since ld-linux.so.2 will interpret the
    filename in the callee's namespace (since removal of "--fd" option
    in 1.13).  No cwd is used for resolving this.

 -- Mark Seaborn <mseaborn@onetel.com>  Tue,  6 Dec 2005 21:21:55 +0000

plash (1.15) unstable; urgency=low

  * Add gtk-powerbox.c:  an LD_PRELOADed patch to Gtk to replace the
    GtkFileChooserDialog interface so that it calls Plash's powerbox.
  * Rename "plash" executable to "pola-shell".
  * fs-operations.c:  Add log method.  Add an "end" log message when
    the fs_op object is dropped.
  * gettextization
  * make.sh:  Add "-Wl,-z,relro" when linking ld.so.  Fixes obscure
    problem when dlopen()ing libraries that might require an executable
    stack.
  * Intercept getsockname() so that it returns the correct pathname for
    Unix domain sockets.
    Extended the g_fds array in libc so that it can contain these
    pathnames.  It is now an array of "struct libc_fd"s, rather than
    an array of "cap_t"s.
    libc-fds.h:  New file.
    libc-misc.c:  Introduced fds_resize(), fds_slot_clear().
    Changed open(), close(), dup2(), etc.
    libc-connect.c:  Add getsockname() and change connect() and bind().
     * Reason:  I discovered that gconfd2 (or possibly Orbit) was
       relying on getsockname() returning the pathname that it earlier
       passed to bind().  This meant that Gnumeric was unable to
       spawn a gconf process itself, and it produced loads of errors.
  * fs-operations.c, libc-misc.c:  Fixed fstat() to return the correct
    information on directory FDs.  Added the fsop_dir_fstat method to
    implement this.
  * build-fs-dynamic.c:  Implement link() and rename() methods.  This
    is needed for when GNOME and KDE apps hard link files inside $HOME.
  * filesysobj-real.c:  Changes to allow rename and hard link calls of
    the form rename("dir/foo1", "dir/foo2") to work.
     * The problem: The real_dir_rename and real_dir_link methods only
       work in the same-directory case; their test was a pointer
       comparison on real_dir objects.  However, resolving a directory
       pathname like "dir" always returns a new real_dir object.  This
       meant that the rename() call wouldn't work when you use full
       pathnames.
     * This was causing some failures.  eg. Konqueror wouldn't start:
       some code relied on creating "$HOME/.ICEauthority-l" as a hard
       link to "$HOME/.ICEauthority-c".
     * The partial solution: Change the same-directory check to
       compare inode and device number of directory, after trying a
       pointer comparison.

 -- Mark Seaborn <mseaborn@onetel.com>  Tue,  6 Dec 2005 21:21:55 +0000

plash (1.14) unstable; urgency=low

  * Add powerbox:  powerbox.[ch], powerbox-req.c, powerbox.el.
    Changed pola-run.c to add --powerbox option etc.
  * cap-protocol.c:  Add support for using glib, so that its event loop
    can be subsumed into glib's event loop, and we can use Gtk.
  * libc-misc.c:  Now intercept the dup2() call.
     * I was going to handle close() and dup2() on the comm socket FD
       by relocating the comm socket to another FD number, but this
       turned out not to be a good idea (so the code is #if'd out).
       cap-protocol.c:  Add cap_relocate_fd().
     * That's because libc.so cannot relocate the FD by itself.  It
       would need to be able to tell ld.so (but it has no way to).
       Really, the application should have control over which FD
       number is used, and I make use of that elsewhere.
     * So instead, dup2() and close() now just refuse the clobber the
       socket FD.
  * Replace use of ld-linux.so.2's --library-path argument with use of
    the LD_LIBRARY_PATH environment variable.  Change run-as-anonymous
    to take extra arguments for setting environment variables.  This is
    needed because run-as-anonymous runs setuid, and this causes the
    dynamic linker to unset LD_LIBRARY_PATH, so this needs to be
    restored.
    Changed shell.c, pola-run.c, fs-operations.c.
  * shell.c:  Change some cases to use format strings -- the beginnings
    of gettext support.
  * pola-run.c:  Now export fs_op_maker object to the application.
  * libc-misc.c:  Change statfs().  Provide a way to fake the amount of
    free space by setting the PLASH_FAKE_STATFS_AVAIL environment
    variable.  This was needed to make the Firefox installer run.


setuid/run-as-anonymous.c
/* Copyright (C) 2004 Mark Seaborn

   This file is part of Plash, the Principle of Least Authority Shell.

   Plash is free software; you can redistribute it and/or modify it
   under the terms of the GNU Lesser General Public License as
   published by the Free Software Foundation; either version 2.1 of
   the License, or (at your option) any later version.

   Plash is distributed in the hope that it will be useful, but
   WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
   Lesser General Public License for more details.

   You should have received a copy of the GNU Lesser General Public
   License along with Plash; if not, write to the Free Software
   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307
   USA.  */

/* This program uses a lock directory to allocate user IDs.
   
   I considered putting the lock directory in /tmp.  We can create a
   directory that only root can read or write.  Other users should not
   be able to rename the directory.
   
   The nice thing about this is that the directory would be cleaned on
   a reboot without having to add any new scripts to the boot
   sequence.
   
   Unfortunately, it allows a possible denial-of-service attack,
   because someone could create a directory in /tmp with the same name
   as the one we're going to use.

   So the directory will have to go somewhere safer in the directory
   tree.  The problem of cleaning the directory is something we have
   to deal with between reboots anyway.

   NB. There's no point in checking the permissions and ownership of
   the directory alone, because if a non-root user owns a directory in
   its path, they can redirect the path to different directories that
   root happens to own using symlinks.
*/

#include <stdlib.h>
#include <alloca.h>
#include <errno.h>
#include <stdio.h>
#include <unistd.h>
#include <grp.h>
#include <sys/time.h>
#include <fcntl.h>
#include <assert.h>
#include <sys/file.h>

#include <string.h>
#include <signal.h>

#include "config.h"


#define NAME "run-as-anonymous"

#define UID_RANGE_LEN	(UID_RANGE_END - UID_RANGE_START)

/* In glibc, setuid() will try both the 16-bit and 32-bit versions of
   the kernel system call, while glibc provides no setuid32() symbol.
   In dietlibc, setuid() is the 16-bit syscall, and setuid32() is the
   32-bit syscall. */
#ifdef USE_DIETLIBC
#define setuid setuid32
#define setgid setgid32
#define getuid getuid32
#define getgid getgid32
#endif


char hex_digit(int x)
{
  if(0 <= x && x <= 9) return x + '0';
  if(0xa <= x && x <= 0xf) return x - 0xa + 'a';
  assert(0);
  return 'X';
}

int pick_uid()
{
  int try_uid;
  int start_uid;
  struct timeval time;
  gettimeofday(&time, 0);

  assert(UID_RANGE_LEN > 0);
  start_uid = UID_RANGE_START + ((time.tv_sec ^ time.tv_usec) % UID_RANGE_LEN);
  try_uid = start_uid;

  while(1) {
    int fd;
    int i;
    char buf[8 + 1];
    for(i = 0; i < 8; i++) {
      buf[i] = hex_digit((try_uid >> ((7-i)*4)) & 0xf);
    }
    buf[8] = 0;

    /* fprintf(stderr, NAME ": trying uid %s\n", buf); */
    fd = open(buf, O_WRONLY | O_CREAT | O_EXCL, 0600);
    if(fd >= 0) {
      close(fd);
      return try_uid;
    }
    if(errno != EEXIST) {
      perror(NAME ": open");
      return 0;
    }
    try_uid++;
    if(try_uid == start_uid) {
      fprintf(stderr, NAME ": out of uids\n");
      return 0;
    }
    if(try_uid >= UID_RANGE_END) try_uid = UID_RANGE_START;
  }
}

int main(int argc, char **argv)
{
  if(argc < 2) {
    fprintf(stderr, "Usage: %s [--debug] [-s ENVVAR=VALUE]... program args...\n", argv[0]);
    return 1;
  }
  else {
    char **argv2;
    int i;
    int uid;
    int lock_file_fd;

#ifdef IN_CHROOT_JAIL
    if(chdir(UID_LOCK_DIR2) < 0) {
      perror(NAME ": chdir: can't change to lock directory");
      return 1;
    }
#else
    if(chdir(UID_LOCK_DIR) < 0) {
      perror(NAME ": chdir: can't change to lock directory");
      return 1;
    }
#endif

    /* Claim non-exclusive lock before continuing. */
    /* There is no race condition between instances of run-as-anonymous
       without this (hence the non-exclusive lock), but there would be a
       race condition with gc-uid-locks.  Between creating the lock file
       for a UID and calling setgid()/setuid(), gc-uid-locks would
       think that the UID is not in use. */
    lock_file_fd = open("flock-file", O_RDONLY);
    if(lock_file_fd < 0) {
      perror(NAME ": open: " UID_LOCK_DIR "/flock-file");
      return 1;
    }
    /* Get a shared lock.  Multiple instances of run-as-anonymous can
       run concurrently with each other, but not concurrently with
       gc-uid-locks. */
    if(flock(lock_file_fd, LOCK_SH) < 0) {
      perror(NAME ": flock");
      return 1;
    }
    
    uid = pick_uid();
    if(uid == 0) return 1;

#if !defined IN_CHROOT_JAIL
    if(chroot(JAIL_DIR) < 0) { perror(NAME ": chroot: " JAIL_DIR); return 1; }
#endif
    if(chdir("/") < 0) { perror(NAME ": chdir: /"); return 1; }

    if(setgroups(0, 0) < 0) { perror(NAME ": setgroups"); return 1; }
    if(setgid(uid) < 0) { perror(NAME ": setgid"); return 1; }
    if(setuid(uid) < 0) { perror(NAME ": setuid"); return 1; }
    if(getuid() != uid) {
      fprintf(stderr, NAME ": uid not set correctly to 0x%x (got 0x%x) -- "
	      "does your kernel support 32-bit uids?\n", uid, getuid());
      return 1;
    }
    if(getgid() != uid) {
      fprintf(stderr, NAME ": gid not set correctly to 0x%x (got 0x%x) -- "
	      "does your kernel support 32-bit uids?\n", uid, getgid());
      return 1;
    }

    if(close(lock_file_fd) < 0) { perror(NAME ": close"); return 1; }

    /* From this point on, the code is running as unprivileged.
       It is not quite so critical, from the point of view of whoever
       installs this as setuid root. */

    argc--;
    argv++;

    /* Process the --debug option.  This suspends the process and
       prints the PID.  This is useful for attaching gdb to the
       process. */
#ifndef NO_DEBUG
    if(argc >= 2 && !strcmp(argv[0], "--debug")) {
      fprintf(stderr, NAME ": stopping, pid %i\n", getpid());
      if(raise(SIGSTOP) < 0) { perror(NAME ": raise(SIGSTOP)"); return 1; }
      
      argc--;
      argv++;
    }
#endif

    /* Process -s options for setting environment variables.  This is
       useful for restoring variables such as LD_LIBRARY_PATH, which
       the dynamic linker will unset when run-as-anonymous starts
       because it notices that the process is setuid. */
    while(argc >= 3 && !strcmp(argv[0], "-s")) {
      if(putenv(argv[1]) < 0) {
	fprintf(stderr, NAME ": putenv failed\n");
	return 1;
      }
      argc -= 2;
      argv += 2;
    }

    argv2 = alloca((argc + 1) * sizeof(char *));
    for(i = 0; i < argc; i++) argv2[i] = argv[i];
    argv2[i] = 0;
    
    execv(argv[0], argv2);
    fprintf(stderr, NAME ": exec: %s: %s\n", argv[0], strerror(errno));
    return 1;
  }
}
